Changing multi-distro vars to facts

macunha1 opened the pull request:
Before jumping into it: really appreciate the work done here together with the GitHub actions pipeline + Molecule action. Thank you for sharing these gems!

Debian doesn't (strictly) depend on the values coming from
https://github.com/ome/ansible-role-postgresql-client/blob/master/vars/main.yml#L5-L21

but it will still fail at

https://github.com/ome/ansible-role-postgresql/blob/ffb2effedc5514bfc0fbac2be72fad223cfa8168/tasks/initialise.yml#L14-L26

since that will try to load default values for RedHat (+ versions).

Setting Postgres versions that aren't previously defined on PostgreSQL client (e.g. 13, 14) are working just fine on Debian after implementing these changes.

To segregate both installations, the values are now set using `set_fact` inside its own distro's tasks file, which corresponds to the conditional on tasks/main.yml and ensures consistency.

Some bonus changes:
+ included a block on tasks to have become+become_user defined less
  times;
+ added a variable to change the user that it will become when asked.

macunha1 committed:
Changing multi-distro vars to facts

Debian doesn't (strictly) depend on the values coming from https://github.com/ome/ansible-role-postgresql-client/blob/master/vars/main.yml#L5-L21
but it will still fail at https://github.com/ome/ansible-role-postgresql/blob/ffb2effedc5514bfc0fbac2be72fad223cfa8168/tasks/initialise.yml#L14-L26
since that will try to load default values from redhat (+ versions).

Setting Postgres versions that aren't previously defined on Debian works
just fine, Postgres 13 and 14 installed without and issue.

To segregate both installations, the values are now set using set_fact
inside its own distro's tasks file, which corresponds to the conditional
on tasks/main.yml and ensures consistency.

Some bonus changes:
+ included a block on tasks to have become+become_user defined less
  times;
+ added a variable to change the user that it will become when asked.

sbesson reviewed (changes requested):
Thanks @macunha1 for your contribution.

The Molecule tests are failing with "lines too long" errors. I suspect this is related to the indentation of the tasks with the `block` addition.

To fix these issues but also help focus the scope of the PR and facilitate its review, would it be possible to remove the last two features (usage of `block` + parametrization of `postgresql_become_user`) and focus on the Debian/RedHat variables first?

macunha1 committed:
style(lint): Adjusting lines that were too big

Broke down lines that are bigger than 80 characters

macunha1 committed:
fix(undefined): Changing service name variable

Adjusting the variable name in handlers/main.yaml to incude the defined
Ansible fact.

macunha1 commented:
thank you for pointing that out @sbesson.

The tests are now passing, let me know in case it is necessary to change something else.

sbesson reviewed (commented):
Overall the proposed changes make sense to me. 

Effectively this PR hides the distribution-specific details from playbooks by splitting the Debian/RedHat into separate internal variables and letting the role filtering these and converting them in unified `postgresql_dist_*` variables.

We have limited experience with the `block` statement but I find this change illustrates well how it can cut a lot of redundancy.

I have two thoughts mostly regarding the API contract:

- are you considering the `postgresql_dist_*` variables as part of the public API of this role i.e. is it something that not only the role but also the playbooks could start depending on? If so, it might be worth documenting e.g. in the README so that we would consider modifications as a breaking change
- also around this private/public delineation, I have been thinking whether we should prefix private variables with `_` using the standard Python style e.g.  `_postgresql_dist_debian`

/cc @joshmoore

macunha1 commented:
> * are you considering the `postgresql_dist_*` variables as part of the public API of this role i.e. is it something that not only the role but also the playbooks could start depending on?

From my perspective, keeping the variables as part of the public API would be preferable to enable customizations by users (when/if one day necessary), it allows the role to be future-proof without requiring constant maintenance. e.g.: a user could parameterize the role to have Pg 13+ running on RedHat.

If that sounds plausible, I would be glad to write the README breaking change instructions.

What do you think?

sbesson commented:
Sorry for the delay and happy new year. Yes please push the related README changes. Also you are mentioning breaking changes: is there anything else than the `postgresql_dist_{debian,redhat}` internal variables?

macunha1 committed:
doc(*): Adding postgresql_dist_<OS> to README

Including a short description about postgresql_dist_redhat and
postgresql_dist_debian

macunha1 commented:
Happy new year! :tada: 

> [...] you are mentioning breaking changes: is there anything else than the `postgresql_dist_{debian,redhat}` internal variables?

Sorry about the misunderstanding, I used the same term (now I realize it was misleading) from https://github.com/ome/ansible-role-postgresql/pull/25#pullrequestreview-831686128

> [...] in the README so that we would consider modifications as a breaking change.

Just pushed some instructions to README + comments on `defaults/main.yml` @sbesson

sbesson commented:
Thanks @macunha1. Looks good and I would propose to get this released as 5.1.0 i.e. a minor version increment. Before merging and tagging this, could you sign and return the OME CLA - see the instructions in the [contributing documentation](https://ome-contributing.readthedocs.io/en/latest/cla.html)?

macunha1 commented:
> Thanks @macunha1. Looks good and I would propose to get this released as 5.1.0 i.e. a minor version increment. Before merging and tagging this, could you sign and return the OME CLA - see the instructions in the [contributing documentation](https://ome-contributing.readthedocs.io/en/latest/cla.html)?

Sorry about the delay, just signed and sent it to the mentioned e-mail.

sbesson commented:
@macunha1 thanks for your contribution! These changes have now been released in a minor release of this role and are available from Galaxy https://galaxy.ansible.com/ome/postgresql
